### PR TITLE
Couchdb 3326 clustered purge

### PR DESCRIPTION
**Pull request by jiangphcn**
## Overview

The primary purpose of clustered purge is to clean databases that have multiple deleted tombstones or single documents that contain large numbers of conflicts. But it can also be used to purge any document (deleted or non-deleted) with any number of revisions.

Clustered Purge is designed to maintain eventual consistency and prevent unnecessary invalidation of secondary indexes. For this, every database keeps track of a certain number of historical purges requested in the database, as well as its current ``purge_seq``. Internal replications and secondary indexes process database's purges and periodically update their corresponding purge checkpoint documents to report ``purge_seq`` processed by them. To ensure eventual consistency, the database will remove store historical purge requests only after they have been processed by internal replication jobs and secondary indexes.

## Testing recommendations

`make html` and `make check`
## GitHub issue number

https://issues.apache.org/jira/browse/COUCHDB-3326

## Related Pull Requests

https://github.com/apache/couchdb/pull/1370

## Checklist

- [X] Documentation is written and is accurate;
- [X] `make check` passes with no errors
- [ ] Update [rebar.config.script](https://github.com/apache/couchdb/blob/master/rebar.config.script) with the commit hash once this PR is rebased and merged

### CODE REVIEW

File: src/api/database/misc.rst [hunk 1]
```diff
@@ -17,40 +17,22 @@
...
-    .. warning::
-        Purging a document from a database should only be done as a last resort
-        when sensitive information has been introduced inadvertently into a
-        database. In clustered or replicated environments it is very difficult
-        to guarantee that a particular purged document has been removed from
-        all replicas. Do not rely on this API as a way of doing secure
-        deletion.
+    The format of the purge request must include the document IDs and for each
+    document ID one or more revisions that must be purged. Documents can be
+    prior deleted, but it is not necessary. Revisions must be leaf revisions.
```
Comment (flimzy) [inline src/api/database/misc.rst]: "prior" should be "previously"

File: src/api/database/misc.rst [hunk 2]
```diff
@@ -17,40 +17,22 @@
...
 
-    .. warning::
-        Purging a document from a database should only be done as a last resort
-        when sensitive information has been introduced inadvertently into a
-        database. In clustered or replicated environments it is very difficult
-        to guarantee that a particular purged document has been removed from
-        all replicas. Do not rely on this API as a way of doing secure
-        deletion.
+    The format of the purge request must include the document IDs and for each
+    document ID one or more revisions that must be purged. Documents can be
```
Comment (flimzy) [inline src/api/database/misc.rst]: Missing some commas. Should be: "...must include the document IDs, and for each document ID, one or more revisions that must be purged."

File: src/api/database/misc.rst [hunk 3]
```diff
@@ -17,40 +17,22 @@
...
     identify that the document has been deleted by its absence.
 
-    .. warning::
-        Purging a document from a database should only be done as a last resort
-        when sensitive information has been introduced inadvertently into a
-        database. In clustered or replicated environments it is very difficult
-        to guarantee that a particular purged document has been removed from
-        all replicas. Do not rely on this API as a way of doing secure
-        deletion.
+    The format of the purge request must include the document IDs and for each
```
Comment (flimzy) [inline src/api/database/misc.rst]: I suggest removing "format" from this sentence. Instead, just: "The purge request must..."

File: src/api/database/misc.rst [hunk 4]
```diff
@@ -59,11 +41,14 @@
     :<json object: Mapping of document ID to list of revisions to purge
     :>header Content-Type: - :mimetype:`application/json`
                            - :mimetype:`text/plain; charset=utf-8`
-    :>json number purge_seq: Purge sequence number
+    :>json string purge_seq: Purge sequence string
     :>json object purged: Mapping of document ID to list of purged revisions
-    :code 200: Request completed successfully
+    :code 201: Request completed successfully
+    :code 202: Request was accepted, and at least completed successfully on one
+                of the replicas, but the quorum was not reached
```
Comment (flimzy) [inline src/api/database/misc.rst]: Suggest rewording, for clarity:  "Request was accepted, and was completed successfully on at least one replica, but quorum was not reached."

File: src/api/database/misc.rst [hunk 5]
```diff
@@ -59,11 +41,14 @@
...
     :code 400: Invalid database name or JSON payload
     :code 415: Bad :header:`Content-Type` value
+    :code 500: Internal server errors or timeout
```
Comment (flimzy) [inline src/api/database/misc.rst]: "Internal server error or timeout"  (not plural)

File: src/api/database/misc.rst [hunk 6]
```diff
@@ -78,47 +63,216 @@
...
-        Content-Length: 103
+        Content-Length: 107
         Content-Type: application/json
-        Date: Mon, 12 Aug 2013 10:53:24 GMT
-        Server: CouchDB (Erlang/OTP)
+        Date: Fri, 02 Jun 2017 18:55:54 GMT
+        Server: CouchDB/2.0.0-2ccd4bf (Erlang OTP/18)
 
         {
-            "purge_seq":3,
```
Comment (flimzy) [inline src/api/database/misc.rst]: Did you intend to eliminate "purge_seq" from this example? The docs still mention it above (but now as a string, not a number)

File: src/api/database/misc.rst [hunk 7]
```diff
@@ -78,47 +63,216 @@
...
+     :alt: Document Revision Tree 1
+
+     Document Revision Tree 1
+
+For example, given the above purge tree and issuing the above purge request,
+the whole document will be purged, as it contains only a single branch with a
+leaf revision  `3-c50a32451890a3f1c3e423334cc92745` that will be purged.
+As a result of this purge operation, a document with
+`_id:c6114c65e295552ab1019e2b046b10e` will be completely removed from the
+database's documents's b+tree, and sequence b+tree. It will not be available
```
Comment (flimzy) [inline src/api/database/misc.rst]: replace "documents's" with "document"

File: src/api/database/misc.rst [hunk 8]
```diff
@@ -78,47 +63,216 @@
...
+through `_all_docs` or `_changes` APIs, as though this document never existed.
```
Comment (flimzy) [inline src/api/database/misc.rst]: suggest replacing "APIs" with "endpoints"

File: src/api/database/misc.rst [hunk 9]
```diff
@@ -78,47 +63,216 @@
...
+======================
+Purges are automatically replicated between replicas of the same database. Each
+database has an internal purge tree that stores a certain number of the most
+recent purges. This allows internal synchonization between replicas of the same
+database.
+
+External Replication
+======================
+Purge operations are not replicated to other external databases. External
+replication works by identifying source's document revisions that are missing on
```
Comment (flimzy) [inline src/api/database/misc.rst]: should be:  "works by identifying a source's document..."

File: src/api/database/misc.rst [hunk 10]
```diff
@@ -78,47 +63,216 @@
...
+target, and copying these revisions from source to target. A purge operation
+completely purges revisions from a document's purge tree making external
+replication of purges not possible.
```
Comment (flimzy) [inline src/api/database/misc.rst]: replace "not possible" with "impossible"

File: src/api/database/misc.rst [hunk 11]
```diff
@@ -78,47 +63,216 @@
...
+
+    .. note::
+      If you are replicating between external databases and have
+      purged any documents and interested in having corresponding
+      purges in other databases as well, you should run purge
+      separately on each of the external databases.
```
Comment (flimzy) [inline src/api/database/misc.rst]: I think this should be re-worded for clarity. I suggest:

> If you need a purge to be effective across multiple effective databases, you must run the purge separately on each of the databases.

Comment (flimzy) [reply src/api/database/misc.rst]: A question here: Is it worth mentioning that these purges should be done while replications are not running? Otherwise, people may end up in a race, purging docs, just to see them recreated during replication.... if I'm not mistaken?

File: src/api/database/misc.rst [hunk 12]
```diff
@@ -78,47 +63,216 @@
...
 purged documents.
 
-When the indexer identifies that the purge sequence on a database has changed,
-it compares the purge sequence of the database with that stored in the view
-index. If the difference between the stored sequence and database is sequence
-is only 1, then the indexer uses a cached list of the most recently purged
-documents, and then removes these documents from the index individually. This
-prevents completely rebuilding the index from scratch.
+Each internal database indexer, including the view indexer, keeps its own purge
+sequence. The purge sequence stored in index can be much smaller than the
```
Comment (flimzy) [inline src/api/database/misc.rst]: should be "in the index"

File: src/api/database/misc.rst [hunk 13]
```diff
@@ -78,47 +63,216 @@
...
+database's purge sequence up to the number of purge requests allowed to be
+stored in the purge trees of the database. Multiple purge requests can be
+processed by indexer without incurring rebuilt of the index. The index will be
```
Comment (flimzy) [inline src/api/database/misc.rst]: "by the indexer"

Comment (flimzy) [reply src/api/database/misc.rst]: replace "rebuilt" with "a rebuild"

File: src/api/database/misc.rst [hunk 14]
```diff
@@ -78,47 +63,216 @@
...
+updated according to these purge requests.
+
+The index of document is based on winner of revision tree. Depending on which
+revision specified in purge request, there are different behaviors to update
+index:
```
Comment (flimzy) [inline src/api/database/misc.rst]: This sentence needs to be re-worded. I suggest:

> The index of documents is based on the winner of the revision tree. Depending on which revision is specified in the purge request, the index update observes the following behavior

File: src/api/database/misc.rst [hunk 15]
```diff
@@ -78,47 +63,216 @@
...
+
+- if winner of revision tree is not specified in purge request, there is no
+  change on index record of this document;
```
Comment (flimzy) [inline src/api/database/misc.rst]: - If the winner of the revision tree is not specified in the purge request, there is no change to the index record of this document.

File: src/api/database/misc.rst [hunk 16]
```diff
@@ -78,47 +63,216 @@
...
+- if winner of revision tree is specified in purge request and there is still
+  revision left after purging, the index record of document will be built
+  according to new winner of revision tree;
```
Comment (flimzy) [inline src/api/database/misc.rst]: - If the winner of the revision tree is specified in the purge request, and there is still a revision left after purging, the index record of the document will be built according to the new winner of the revision tree.

File: src/api/database/misc.rst [hunk 17]
```diff
@@ -78,47 +63,216 @@
...
+- if all revisions of document are specified in purge request, the index record
+  of document will be deleted. The document can't be searched out any more.
```
Comment (flimzy) [inline src/api/database/misc.rst]: - If all revisions of the document are specified in the purge request, the index record of the document will be deleted. The document will no longer be found in searches.

File: src/api/database/misc.rst [hunk 18]
```diff
@@ -78,47 +63,216 @@
...
+
+.. _api/db/_purged_infos_limit:
+
+==============================
+``/db/_purged_infos_limit``
+==============================
+
+.. http:get:: /{db}/_purged_infos_limit
```
Comment (flimzy) [inline src/api/database/misc.rst]: I know this is only a documentation PR, but if it's not too late, can this endpoint be renamed?  "infos" sounds very unnatural--we never speak about the plural of info in English. I'd suggest "_purged_docs_limit", since that's what it's described as below.  :)

File: src/api/database/misc.rst [hunk 19]
```diff
@@ -78,47 +63,216 @@
...
+    :synopsis: Returns the limit of historical purges to store in the database
+
+    Gets the current ``purged_infos_limit`` (purged documents limit) setting,
+    the maximum number of historic purges (purged document Ids with their
```
Comment (flimzy) [inline src/api/database/misc.rst]: replace "historic" with "historical"

File: src/api/database/misc.rst [hunk 20]
```diff
@@ -78,47 +63,216 @@
...
+
+    The default value of historical stored purges is 1000. This means up to 1000
+    purges can be synchronized between replicas of the same databases in case of
+    one of the replicas was down when purges occurred.
+
+    This request sets the soft limit for stored purges. During the compaction
+    CouchDB will try to keep only `_purged_infos_limit` of purges in the
+    database, but occasionally the number of stored purges can exceed this
+    value. If a database has not completed purge synchronization with active
+    indexes or active internal replications, it may temporarily store higher
```
Comment (flimzy) [inline src/api/database/misc.rst]: should be "a higher"

File: src/cluster/purging.rst [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+===============
+Clustered Purge
+===============
+The primary purpose of clustered purge is to clean databases that have multiple
+deleted tombstones or single documents that contain large numbers of conflicts.
+But it can also be used to purge any document (deleted or non-deleted) with any
+number of revisions.
+
+Clustered Purge is designed to maintain eventual consistency and prevent
```
Comment (flimzy) [inline src/cluster/purging.rst]: "Purge" should be made lowercase:  "Clustered purge is designed..."

File: src/cluster/purging.rst [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+unnecessary invalidation of secondary indexes. For this, every database keeps
+track of a certain number of historical purges requested in the database, as
+well as its current ``purge_seq``. Internal replications and secondary indexes
+process database's purges and periodically update their corresponding purge
+checkpoint documents to report ``purge_seq`` processed by them. To ensure
+eventual consistency, the database will remove store historical purge requests
```
Comment (flimzy) [inline src/cluster/purging.rst]: remove the word "store":  "the database will remove historical purge requests..."

Or maybe it should be "stored"?  "the database will remove stored historical purge requests..."

File: src/cluster/purging.rst [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+
+Compaction of Purges
+====================================
+
+During the compaction of the database the oldest purge requests are to be
+removed to store only ``purged_infos_limit`` number of purges in the database.
+But in order to keep the database consistent with indexes and other replicas,
+we can only remove purge requests that have already been processed by indexes
+and internal replications jobs. Thus, occasionally purge trees may store
+purges in the number greater than ``purged_infos_limit``. If the number of
```
Comment (flimzy) [inline src/cluster/purging.rst]: I suggest rewording this sentence for clarity:

> Thus, occasionally purge trees may store more than ``purged_infos_limit`` purges.

File: src/cluster/purging.rst [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+Purge requests are replayed across all nodes in an eventually consistent manner.
+Internal replication of purges consists of two steps:
+1. Pull replication. Internal replication first starts by pulling purges from
+target and applying them on source to make sure we don't reintroduce to target
+source's docs/revs that have been already purged on target. In this step, we use
+purge checkpoint documents stored on target to keep track of the last target's
+``purge_seq`` processed by the source. We find purge requests occurred after
+this ``purge_seq``, and replay them on source. This step is done by updating
+target's checkpoint purge documents with the latest process ``purge_seq`` and
+timestamp.
```
Comment (flimzy) [inline src/cluster/purging.rst]: "by updating the target's"

File: src/cluster/purging.rst [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+2. Push replication. Then internal replication proceeds as usual with an extra
+step inserted to push source's purge requests to target. In this step, we use
+local internal replication checkpoint documents, that are updated both on target
+and source.
+
+Under normal conditions, a interactive purge request is already sent to every
```
Comment (flimzy) [inline src/cluster/purging.rst]: replace "a" with "an":  "an interactive purge"...

File: src/cluster/purging.rst [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+node containing database shard's replica, and applied on every replica. Internal
```
Comment (flimzy) [inline src/cluster/purging.rst]: "a database shard's"

File: src/cluster/purging.rst [hunk 27]
```diff
@@ -147,7 +147,7 @@ These settings ca be updated in the default.ini or local.ini:
         During a database compaction,  we check all checkpoint purge docs. A client (an
         index or internal replication job) is allowed to have the last reported
         ``purge_seq`` to be smaller than the current database shard's ``purge_seq`` by
-        the value of ``(purged_docs_limit + index_lag_warn_seconds)``.  If the client's
+        the value of ``(purged_infos_limit + index_lag_warn_seconds)``.  If the client's
```
Comment (flimzy) [inline src/cluster/purging.rst]: I'm glad this change was possible! :D

File: src/api/database/misc.rst (hunk 18 above)
Comment (jiangphcn) [reply src/api/database/misc.rst]: We changed from `_purged_docs_limit` to `_purged_infos_limit`. Let us check with Paul @davisp about his preference.

**Comment by jiangphcn** [comment]
Thanks so much @flimzy for your valuable comments. Fixing them makes this document better. Appreciate it. Regarding endpoint name, let us check with Paul again.

File: src/api/database/misc.rst (hunk 6 above)
Comment (jiangphcn) [reply src/api/database/misc.rst]: added again :-)

**Comment by davisp** [comment]
+1 Good write up

**Comment by jiangphcn** [comment]
okay @wohali hold off a while
